### PR TITLE
Pixtral: vectorize patch embeddings and enable tests

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Continuation on the discussion from https://github.com/huggingface/transformers/pull/35110.

This PR gets rid of the loop over each image in the input for Pixtra and makes it more aligned with other VLMs. Now the model will pad the image on h/w dimensions and unpad it back after the vision patch embedding layer. That also helps us to get rid of extra dimension errors on processing code we've been having lately and remove the `BacthFeatureMix`

Test with the demo script from https://huggingface.co/mistral-community/pixtral-12b, as we don't have any test for pixtral as VLM at the moment. The generations match on text level

cc @Rocketknight1 wdyt about this? The design might need some changes as I had to make llava accept extra kwargs (`image_sizes`) to make the model work

### CODE REVIEW

**Comment by ywang96** [comment]
@zucchini-nlp One concern I have about `padding the images before the vision backbone and then unpadding them back after` is that then you'll be always padding every image in the batch to the max size unnecessarily (especially when models like Pixtral support 10+ multi-image in one inference call), which IMO could be a bit problematic. At that point it maybe just better to flatten and concat pixel values to the batched tensor of shape (total_num_patches, C, H, W) since you will need to pass in additional information (image_sizes in this case) anyways.

**Comment by zucchini-nlp** [comment]
Made some evals for memory usage with inputs where all images are low resolution (100 x 100) and there is one with the highest resolution possible (1024 x 1024). The memory usage between old and the new processing logic is almost same, excluding memory used by model weights it is 

Old-processor -> 41470 MiB
New processor -> 41875 MiB

```python
from PIL import Image
import torch
from transformers import AutoProcessor, LlavaForConditionalGeneration

model_id = "mistral-community/pixtral-12b"
processor = AutoProcessor.from_pretrained(model_id, use_fast=True)
processor.tokenizer.pad_token_id = processor.tokenizer.eos_token_id
processor.tokenizer.padding_side = "left"


url_dog = "https://picsum.photos/id/237/100/100"
url_mountain = "https://picsum.photos/seed/picsum/100/100"
url_stop = "https://picsum.photos/id/237/2000/2000"

chat = [
    {
      "role": "user", "content": [
        {"type": "text", "content": "Can this animal"}, 
        {"type": "image"},
        {"type": "text", "content": " live here?"}, 
        {"type": "image"},
      ]
    }
]
prompt = processor.apply_chat_template(chat)

chat_2 = [
    {
      "role": "user", "content": [
        {"type": "text", "content": "What do you see here"}, 
        {"type": "image"},
      ]
    }
]
prompt_2 = processor.apply_chat_template(chat_2)


begin = torch.cuda.memory_allocated()

prompts = [prompt] * 35 + [prompt_2]
images = [[url_dog, url_mountain]] * 35 + [[url_stop]]


inputs = processor(text=prompts, images=images, padding=True, return_tensors="pt")

model = LlavaForConditionalGeneration.from_pretrained(model_id, device_map="cuda:0", torch_dtype="float16")

inputs = inputs.to(model.device, torch.float16)
generate_ids = model.generate(**inputs, max_new_tokens=50)
output = processor.batch_decode(generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False)[0]

end = torch.cuda.max_memory_allocated()
print((end - begin) // 1024 ** 2, "MiB")
```

**Comment by zucchini-nlp** [comment]
Ready for review! Quick update:

This PR modifies Pixtral model code to act similarly to other VLMs that divide each image into patches. So, now we pad the inputs to `max_patch_size` and will have a more standard 4D `pixel_values`. After running the patch embeddings, the `hidden_states` can be unpadded back. Similar method is used in llava-next based models and in Emu3 (maybe also Molmo 🤔 )

Plus, I enabled the tests that were skipped and fixed what needed to be fixed. All slow tests for Pixtral are passing + ran a sanity check by generating with "pixtral-12b"

**Review by Rocketknight1 — APPROVED** [review]
If tests pass, I'm happy with it! In general I strongly approve of making Pixtral more compliant with the other VLM code. Also, the existing code doesn't work with batched inputs at all, so more or less anything is an improvement :sweat_smile:

File: src/transformers/models/pixtral/image_processing_pixtral.py [hunk 1]
```diff
@@ -368,6 +272,49 @@ def resize(
             **kwargs,
         )
 
+    def _pad_for_batching(
+        self,
+        pixel_values: List[np.ndarray],
+        image_sizes: List[List[int]],
+        data_format: Optional[Union[str, ChannelDimension]] = None,
+        input_data_format: Optional[Union[str, ChannelDimension]] = None,
+    ):
```
Comment (ArthurZucker) [inline src/transformers/models/pixtral/image_processing_pixtral.py:281]: btw this is not "official" as I am pretty sure the original model does not support this so let's make sure we properly test!

File: src/transformers/models/pixtral/image_processing_pixtral.py [hunk 2]
```diff
@@ -469,57 +416,59 @@ def preprocess(
...
+                image = self.normalize(
+                    image=image, mean=image_mean, std=image_std, input_data_format=input_data_format
+                )
+
+            batch_images.append(image)
+            batch_image_sizes.append(get_image_size(image, input_data_format))
+
+        batch_images = [
+            to_channel_dimension_format(image, data_format, input_channel_dim=input_data_format)
+            for image in batch_images
```
Comment (ArthurZucker) [inline src/transformers/models/pixtral/image_processing_pixtral.py:459]: why do we have to do this outside the forward loop?

File: src/transformers/models/pixtral/image_processing_pixtral_fast.py [hunk 3]
```diff
@@ -291,64 +320,63 @@ def preprocess(
         )
 
         if do_convert_rgb:
```
Comment (ArthurZucker) [inline src/transformers/models/pixtral/image_processing_pixtral_fast.py:322]: while we are at it, we should ahve a single for loop. cc @yonigozlan let's be careful with this please, but here, yes for a single type you can have a different branch, but here you iterate over the images for: convert to rgb, then for the device, then for the permute, and then one more time for the other processing

File: src/transformers/models/pixtral/modeling_pixtral.py [hunk 4]
```diff
@@ -399,10 +398,9 @@ def forward(
 class PixtralPreTrainedModel(PreTrainedModel):
     config_class = PixtralVisionConfig
     base_model_prefix = "model"
+    main_input_name = "pixel_values"
     supports_gradient_checkpointing = True
-    _no_split_modules = ["PixtralVisionAttention"]
-    _skip_keys_device_placement = "past_key_values"
```
Comment (ArthurZucker) [inline src/transformers/models/pixtral/modeling_pixtral.py:404]: why don't we keep `_skip_keys_device_placement`?

File: src/transformers/models/pixtral/modeling_pixtral.py [hunk 5]
```diff
@@ -399,10 +398,9 @@ def forward(
...
-    _supports_cache_class = True
```
Comment (ArthurZucker) [inline src/transformers/models/pixtral/modeling_pixtral.py:405]: pretty sure it does support cache class

File: src/transformers/models/pixtral/modeling_pixtral.py [hunk 6]
```diff
@@ -490,24 +497,36 @@ def forward(
...
-        patch_embeds_list = [self.patch_conv(img.to(self.dtype)) for sample in pixel_values for img in sample]
+        patch_embeds = self.patch_conv(pixel_values)
+        patch_embeds_list = [
+            embed[..., : (size[0] // self.patch_size), : (size[1] // self.patch_size)]
+            for embed, size in zip(patch_embeds, image_sizes)
+        ]
 
         # flatten to a single sequence
-        patch_embeds = torch.cat([p.flatten(1).T for p in patch_embeds_list], dim=0).unsqueeze(0)
+        patch_embeds = torch.cat([p.unsqueeze(0).flatten(2).permute(0, 2, 1) for p in patch_embeds_list], dim=1)
```
Comment (ArthurZucker) [inline src/transformers/models/pixtral/modeling_pixtral.py:507]: can't we do this with the previous forward loop as well?

**Review by ArthurZucker — COMMENTED** [review]
Thanks! would be nice to add an explicit test with padded patches and etc

File: src/transformers/models/pixtral/modeling_pixtral.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/models/pixtral/modeling_pixtral.py:404]: Oh, it's because Pixtral is a vision-only model, as ViT. And we dont expect ViT to have cache at all. The VLM is anyway called with llava model class, which should be the one showing if we support cache or not (depending on LM)

File: src/transformers/models/pixtral/image_processing_pixtral_fast.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/models/pixtral/image_processing_pixtral_fast.py:322]: yeah sure, let's remove it since we already touched the looping on model-side

File: src/transformers/models/pixtral/image_processing_pixtral.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/models/pixtral/image_processing_pixtral.py:459]: hehe, right, this is the way it was before but now there should be no need

File: src/transformers/models/pixtral/image_processing_pixtral.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/pixtral/image_processing_pixtral.py:281]: Updated the pixtral slow test to use the correct checkpoint, and load in 4bit to fit into T4 runners. Plus added some more tests on processor side for nested/flat list of images as inputs. Should be enough I guess

File: tests/models/llava/test_modeling_llava.py [hunk 7]
```diff
@@ -583,46 +583,48 @@ def test_generation_siglip_backbone(self):
...
-3. **Image 3:**
-   - **Description:** A beach scene with waves crashing against the shore. There are several people in the water and on the beach, enjoying the waves and the sunset.
-   - **Details:** The waves are powerful, creating a dynamic and lively atmosphere. The sky is painted with hues of orange and pink from the setting sun, adding a warm glow to the scene.
-
-4. **Image 4:**
-   - **Description:** A garden path leading to a large tree with a bench underneath it. The path is bordered by well-maintained grass and flowers.
-   - **Details:** The path is made of small stones or gravel, and the tree provides a shaded area with the bench invitingly placed beneath it. The surrounding area is lush and green, suggesting a well-kept garden.
-
-Each image captures a different scene, from a close-up of a dog to expansive natural landscapes, showcasing various elements of nature and human interaction with it.
-"""
```
Comment (ArthurZucker) [inline tests/models/llava/test_modeling_llava.py:625]: IMO this test is important: it makes sure a single prompt can describe 4 images and this test was 1-1 passing so let's keep it

**Review by ArthurZucker — APPROVED** [review]
Thanks! Let's run these slow test and merge

**Comment by ArthurZucker** [comment]
run slow pixtral

File: tests/models/llava/test_modeling_llava.py (hunk 7 above)
Comment (zucchini-nlp) [reply tests/models/llava/test_modeling_llava.py]: oke, but the test wasn't even loading the model correctly for me, so I remove it. I changed the ckpt now and used correct device for inputs

One thing is that the model runs on cpu, we can't put it to gpu even in fp16, and 4-bit hurts performance a lot. Will just leave as is then
